File: 0383-ransom-note/0383-ransom-note.dart
class Solution {
  bool canConstruct(String ransomNote, String magazine) {
    var map_m = {};
    for(var c in magazine.split('')) {
        if (map_m.containsKey(c)) {
            map_m[c] += 1;
        } else {
            map_m[c] = 1;
        }
    }
    for (var r in ransomNote.split('')) {
        if (!map_m.containsKey(r)) {
            return false;
        } else {
            map_m[r] -= 1;
            if (map_m[r] < 0) {
                return false;
            }
        }
    }
    return true;
  }
}

File: 0383-ransom-note/0383-ransom-note.py
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        dict_r = {}
        dict_m = {}
        for i in ransomNote:
            if i in dict_r:
                dict_r[i] += 1
            else:
                dict_r[i] = 1

        for i in magazine:
            if i in dict_m:
                dict_m[i] += 1
            else:
                dict_m[i] = 1

        for k, v in dict_r.items():
            if k not in dict_m:
                return False
            if v > dict_m[k]:
                return False
        else:
            return True